ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/UtxoDiff.hs

```diff
@@ -0,0 +1,177 @@
+     UtxoDiff
+        -- all deletes except the ones cancelling m2
+        (Map.union m1 (m3 `Map.difference` m2))
+        -- inserts from the first component without the latest deletes and inserts
+        -- from the second component
+        (Map.unionWith const (m2 `Map.difference` m3) m4)
+
+instance Ord k => Monoid (UtxoDiff k v) where
+  mempty :: UtxoDiff k v
+  mempty = UtxoDiff mempty mempty
+
+{------------------------------------------------------------------------------
+  Conversion
+------------------------------------------------------------------------------}
+
+keysSet :: Ord k => UtxoDiff k v -> Set k
+keysSet (UtxoDiff m1 m2) = Map.keysSet m1 <> Map.keysSet m2
+
+{------------------------------------------------------------------------------
+  Construction
+------------------------------------------------------------------------------}
+
+diff :: Map k v -> Map k v -> UtxoDiff k v
+diff = UtxoDiff
+
+fromMapInserts :: Ord k => Map k v -> UtxoDiff k v
+fromMapInserts = UtxoDiff mempty
+
+fromMapDeletes :: Ord k => Map k v -> UtxoDiff k v
+fromMapDeletes = flip UtxoDiff mempty
+
+fromListInserts :: Ord k => [(k, v)] -> UtxoDiff k v
+fromListInserts = UtxoDiff mempty . Map.fromList
+
+fromListDeletes :: Ord k => [(k, v)] -> UtxoDiff k v
+fromListDeletes = flip UtxoDiff mempty . Map.fromList
+
+{------------------------------------------------------------------------------
+  Query
+------------------------------------------------------------------------------}
+
+null :: UtxoDiff k v -> Bool
+null (UtxoDiff m1 m2) = Map.null m1 && Map.null m2
+
+size :: UtxoDiff k v -> Int
+size (UtxoDiff m1 m2) = Map.size m1 + Map.size m2
+
+numInserts :: UtxoDiff k v -> Int
+numInserts (UtxoDiff _ m2) = Map.size m2
+
+numDeletes :: UtxoDiff k v -> Int
+numDeletes (UtxoDiff m1 _) = Map.size m1
+
+{------------------------------------------------------------------------------
+  Applying diffs
+------------------------------------------------------------------------------}
+
+applyUtxoDiff ::
+     Ord k
+  => Map k v
+  -> UtxoDiff k v
+  -> Map k v
+applyUtxoDiff m (UtxoDiff deletes inserts) =
+    (m `Map.difference` deletes) `Map.union` inserts
+
+applyUtxoDiffForKeys ::
+     Ord k
+  => Map k v
+  -> Set k
+  -> UtxoDiff k v
+  -> Map k v
+applyUtxoDiffForKeys m ks (UtxoDiff deletes inserts) =
+  applyUtxoDiff
+    m
+    (UtxoDiff (deletes `Map.restrictKeys` (Map.keysSet m `Set.union` ks))
+          (inserts `Map.restrictKeys` (Map.keysSet m `Set.union` ks)))
+
+{-------------------------------------------------------------------------------
+  Filter
+-------------------------------------------------------------------------------}
+
+filterOnlyKey :: (k -> Bool) -> UtxoDiff k v -> UtxoDiff k v
+filterOnlyKey f (UtxoDiff m1 m2) =
+  UtxoDiff (Map.filterWithKey (const . f) m1) (Map.filterWithKey (const . f) m2)
+
+{-------------------------------------------------------------------------------
+  Reductive instances
```

Comment:
The correctness of these instances relies on a relation between the two argument `UtxoDiffs`, so *technically* these instances are unlawful. My main argument against providing unlawful instances is that it introduces (unforeseen) ways in which code can go wrong, and that it will be hard to debug when it does. Having fewer assumptions, or assumptions that we can statically guarantee is probably nicest

Some options:
* Clearly document the assumptions we make, and possibly let the instances thrown an error when the UTXO property is violated.
* Don't provide instance for `UtxoDiff k v`, but rather something like `Maybe (UtxoDiff k v)`, and then the UTXO property violation error case is represented as `Nothing` (and the using code can act on the error)
* (my preference): stop using `fingertree-rm`, and use a normal finger tree instead, which just requires `Monoid` and `Semigroup` instances (which are lawful)

Reply:
Done, with something isomorphic to `Maybe`